### ISSUE TITLE
Understanding Operators and Comparators

### ISSUE THREAD

**Issue by idanh** [facebookresearch/PyTorch-BigGraph#102]
Hi,
I've went over documentation and found it to be insufficient to me for fully understand (intuitively) the way big graph works.

When assigning an operator to a relation, we act on the RHS entity with some function, say with a translation operator. I understand a translation matrix is learned during the training step? Could you please help me understand (1) how exactly is training achieving this? (2) what is the intuition on when to use translation operator vs (say) affine? (3) is the matrix learned a global one and will be the same for all entities?

Regrading comparators, I understand that after training, we can use (for example) a dot comparator to compare between 2 vectors (their dot product), using .prepare() we're also able to normalize (if needed). Please help me understand: (1) In configuration, why is it possible to define a comparator given that we use it after training? (do we use the comparator in the optimisation process?) (2) what is the reason for adding negative vectors when comparing between 2 vectors? (comparator takes 4 arguments, +LS +RS -LS -RS right? I'm referring to the negative arguments)

I am sure that my understanding is lacking and therefor am willing to go over the documents again. Given that I did that a couple of times already and still at lost, I'm asking here.
Thanks.

**Comment by lerks** [comment]
Sorry our documentation wasn't clear enough. Let me try to explain better the points you didn't understand. 

First of all, it's a translation vector, not a matrix (it's one-dimensional). At the beginning of training we create one such vector for each _relation type_ (thus it's not global, nor entity-specific: the same entity may be passed translated using different vectors if it has edges of different types). We initialize each vector with all zeros (so that at the beginning it's a no-op). These vectors are then used in the computation of the loss (in the way you said: they are added to the embeddings of the RHS entities), thus when we do the backwards step we end up computing a gradient for the translation vectors too. Which means that during the optimization step, we also do a step of gradient descent on these vectors. This is how they get trained. (In a distributed setting, since these vectors are used at the same time by all the machines, they are synchronized by the parameter server).

The decision of whether to use translation, complex diagonal, or any other operator is a research question: it is up to the researcher to try to come up with different models, that make intuitive sense to them, and evaluate them on their dataset of interest to see how they perform and then pick the best. [In our documentation](https://torchbiggraph.readthedocs.io/en/latest/related.html) we list some models that were proposed and evaluated in the literature. These are known to perform well, but probably some even better model may be developed one day.

As for comparators, they are used during training too. Training consists in (1) calculating the score predictions for a bunch of edges and then (2) measuring how different those scores are from what we want them to be. This "measure" of distance (in (2)) is the loss function, and by optimizing we are changing our model parameters so that at the next round their predictions will hopefully be closer to the intended values. However, when we calculate the score (for (1)), we do it using a comparator.

If during training we only calculated the scores of positive edges, and calculated the loss on them, then there would be a trivial perfect model: all the entities having the same embedding. This way the score for each edge would be the highest possible, since the embeddings are as close as they can be. To avoid this we need to counterbalance the "attractive" force of positive edges by also computing the score of negative edges which "push away" their two embeddings.

I hope this clarifies some of your doubts. If this helped, it would be really cool if you could go over the documentation, improve it where you think it was lacking and send a pull request. This will help future users who may be in your same situation!

**Comment by idanh** [comment]
@lerks, Thank you very much for your answer.

I think my misunderstanding stems from not having intuition on what operators like translation and linear (which I assume is scaling) do on the embedding space.
Indeed, I went over documentation and for the TransE example you've used translation operator, what would be the effect on that specific graph?

Thanks for explaining the role of comparators at training and where the negative sampling arises from and how it relates to comparators. It make sense to me. 
One question on that if I may, is after training is done, is there a use to passing -LR -LH (negative sides) arguments to a DotComparator while comparing between two vectors (for example)?

> If this helped, it would be really cool if you could go over the documentation, improve it where you think it was lacking and send a pull request. This will help future users who may be in your same situation!

After I have a better understanding of the overall model and system I will sure to contribute back. Thank you!

**Comment by lerks** [comment]
I don't think I really have a great intuition on "how" operators transform the embedding space, but I'll try to share a few thoughts. Let's consider the diagonal operator, instead of the translation one, because it serves better as an example. This operator's parameters are also a 1-D vector of the same dimension as the embeddings, and when applied to an embedding vector it multiplies each of its components pointwise by the corresponding parameter (i.e., it does the Hadamard product of the two).

Suppose that, after you've finished training on some graph, you want to use the learned embeddings as "feature vectors" of the entities for some downstream task. (Let me point out that, although this _usually_ ends up working well, PBG's training process is _not_ optimizing for such a use of the embeddings: the _explicit_ goal of training is to improve the accuracy of score predictions for edges, i.e., how likely is there to be an edge between A and B). If, for example, your dataset is a knowledge base, you are probably hoping that the training process manages to extract from the graph and encode into the embedding information such as whether an entity is a person, an object, a place, ... what period in time it has existed, and so on. You may even hope that each of these characteristics is captured by a different component, say, the first component quantifies how likely the entity is to be a person, the second how much it's likely to be an object, etc. (This is generally _not_ the case: since most models in PBG are invariant by rotation such information is typically spread across many components).

Now back to the operators: if _each_ relation type uses a _different_ operator, then _each_ relation type is able to learn how "relevant" each of the characteristics of the entities is for edges of _that_ type. For example, for the relation type "is capital of", the characteristic of being a person is irrelevant and having it play a role in the calculation may just lead to noise in the score. Therefore, the operator  of the "is capital of" relation type may tune its parameters so that the first component of the parameter vector is zero, which means that the first component of the embedding vectors is multiplied by zero and thus ignored (that component, we said, is the one measuring how much an entity is likely is a person).

The above example (setting a parameter to zero in order to ignore a component) only works if one uses the diagonal operator and the dot product or the cosine distance. In fact, usually, the choices of operators and comparators (and perhaps even loss functions) are not independent: they go hand in hand. Some operators "work well" (and are able to learn something meaningful) only when paired with the right comparator. As a naive example, I imagine that the translation operator works better with an L2 comparator (because they both involve Euclidean distances), whereas the diagonal operator works better with the dot product (because they both involve component-wise multiplications).

It would of course be great to have a mathematically-solid theory of why some operators and comparators perform better than others, and how to choose them best, but for now the state of the research (as I know it) is mostly an experimental search, i.e., coming up with new models and evaluating them.

**Comment by lerks** [comment]
Oh and, yes, after training is complete, if you just want to get the scores for some edges, you don't have to pass any negatives to the comparator. It may however still come in handy depending on _which_ scores you want to calculate for your edges (which are in fact pairs of entities):
- if you want pairwise scores ("one-to-one", the i-th entity on one side with the i-th entity on the other side), then you should pass them as positives;
- if you want the scores for all pairs ("all-to-all", each entity on one side with all the entities on the other side), then you should pass one side as positive and the other as negative.

**Comment by idanh** [comment]
@lerks This solidify my understanding, thank you for your comprehensive answers and your time. I hope you don't mind I follow up, as I'm genuinely infestered to understanding BigGraph framework.

Regrading comparators, I understand the more "trivial" case of one-to-one. You said:
> if you want the scores for all pairs ("all-to-all", each entity on one side with all the entities on the other side), then you should pass one side as positive and the other as negative.

I read this as: comparing all embeddings to all the others. Do you mind pointing me to an example on how that's done (technically, in code) and what is the interpretation of the score when doing that?

Regrading the operators, I want to try and be more specific:
Given an unweighted (all weights equal to 1), directed graph G (of no particular type, can be a social graph), with one relation between vertices (so an edge from v to e specify exactly one relation), configured as follows:

relation operator: translation
comparator: dot
loss_fn: logistic

And given a task to predict missing links between vertices:
(1) Train on G to product embeddings
(2) Use a dot comparator, passing two vectors each time
The higher the score between the two embeddings would (implicitly?) means there should be an edge between them.

My choice of a dot comparators is because the pip distribution (i.e pip install torchbiggraph) doesn't include the l2/squared_l2 yet.

Are those choices sensible?
Thanks again, this thread has been very helpful for me.

**Comment by lerks** [comment]
I guess I made it sound more sophisticated than it actually is. What I meant is this: suppose you have two sets of entities, A_1, ..., A_N and B_1, ..., B_N. There's two way to compare them:
- Compare A_i with B_i, thus A_1 with B_1 only, A_2 with B_2 only, and so on. This gives you a set of scores that is a one-dimensional vector with N entries.
- Compare each A_i with each B_j, thus A_1 is compared with all B_1, ..., B_N, and also A_2 is compared with all B_1, ..., B_N, and so on. Thus the scores are a NxN matrix.

When you call a comparator with LHS+, RHS+, LHS- and RHS-, it computes:
- the 1-to-1 scores between LHS+ and RHS+
- the N-to-N scores between LHS- and RHS+
- the N-to-N scores between LHS+ and RHS-

This is because for each score of a positive edge, we want many negative scores, obtained either by swapping the left- or the right-hand side with another entity.

(In fact comparators do the above calculation simultaneously for many batches, thus it looks slightly different, but this is the gist).

I don't think I can advise you on your choice of model. You said your graph is "of no particular type" but I don't think there's such a thing, or at least I don't think there is one model that works well on every type of graph. What I suggest you do is set up an experiment (basically, split your graph into train and test sets), train multiple models on them and evaluate their performances, so that you can choose the best. You could start with the models we propose in the doc, they are good baselines.

**Comment by idanh** [comment]
@lerks Rereading my comment I see the way I phrased it wasn't the best. In:
> directed graph G (of no particular type, can be a social graph)

I actually meant:
> for example, suppose G is a social graph (my idea was something similar to facebook's followers graph)

What I do now is experimenting with it, but I get some odd results, while I get a high AUC and low pos rank metrics, when comparing the scores of positive and negative edges using DotComparator  and passing v1, LHS+ RHS+ (even if the edge is in fact negative) and the zero vectors of same dimensions for LHS- RHS-, I sometimes see higher scores for negative edges than positive. 

This confuses me, as I expect to get lower scores for those negative edges. I do understand that it might be that there's a missing edge there and I can use this score for link suggestions (i.e. you should follow this person) but I feel like I'm doing something incorrectly, as I'd expect most of negative edges have much lower score distribution than positive.

I've tried several configurations so far so I'm feeling the way I use the comparators could be the issue, or the way I interpret the scores.

Thank you again!

**Comment by lerks** [comment]
So, these issues could be due to several factors, ranging from misconfiguration or misuse of PBG to "design" issues in the experiment. I can help with the former but, unfortunately, I can't provide support for the latter. (In other words, I'm here to do "tech support", not broader "consulting"). However, I can give some pointers to things to check.

First of all, if you're using the dot product and putting zero vectors as negatives, you should be seeing a score of zero (at least when the negative is on the left-hand side, as on the right-hand one you could have an operator that's making it non-zero). Is that the case? If not, then you may be using the comparators incorrectly.

Second, you seem to be doing manual evaluations (you're saying "I sometimes see"). This may give you an inaccurate perception of what's going on. A systematic evaluation using a held-out set of edges is the more reliable way of getting an objective sense of the quality of training, you should do that.

Then, you need to make sure that what you're checking at evaluation time is indeed the goal you're optimizing for at training (i.e., the loss function). Here you seem to be comparing positive and negative scores, so in some sense "ranking" them against each other, so you should be using the ranking loss. If, instead, you want to look at a score as an "absolute" measure of the likelihood of an edge's existence, you may want to check out the logistic loss instead.

More broadly, make sure your data has actually some information that can be learned: we've seen users trying to use PBG for very small graphs (the "B" in the acronym stands for "big"), where there's not much statistical information that can be extracted. More than the number of nodes, what's important is that each node is incident to enough edges to have a good number of "data points" you can learn on. We typically suggest a degree of at least 50.

Also, depending on your graph's size, you may want to tune the dimension, as a higher dimension will capture more information and may overfit on graphs that don't hold enough information (vice versa, small dimensions may underfit on complicated graphs). This is a hyperparameter that you need to tune, using the train+eval approach I mentioned earlier.

This is all that comes to mind. If you're still having issues you may want to post your config so that we can see if anything looks odd there.

**Comment by idanh** [comment]
> I can help with the former but, unfortunately, I can't provide support for the latter.
Fully understandable. Thank you.

> if you're using the dot product and putting zero vectors as negatives, you should be seeing a score of zero
Referring to this: https://github.com/facebookresearch/PyTorch-BigGraph/blob/master/docs/source/downstream_tasks.rst#predicting-the-score-of-an-edge

My code to compare 2 vectors looks like that:
`
scores_1, _, _ = comparator(
    comparator.prepare(torch.tensor(src_embedding.reshape([1,1,d]))),
    comparator.prepare(torch.tensor(dest_embedding.reshape([1,1,d]))),
    torch.empty(1, 0, d),  # Left-hand side negatives, not needed
    torch.empty(1, 0, d),  # Right-hand side negatives, not needed
)
`

> Second, you seem to be doing manual evaluations (you're saying "I sometimes see")
I agree. I should have been more clear about my pipeline:
1. train-test split while keeping the graph connected as much as possible
2. train on train dataset
3. validating on:
`
from torchbiggraph.eval import do_eval
do_eval(val_config)
`

Yields:
`
Stats: pos_rank:  1.64294 , mrr:  0.885424 , r1:  0.807166 , r10:  0.989694 , r50:  0.997303 , auc:  0.987671 , count:  5191
`

Than, I went ahead and anecdotally looked at some examples (>> 10) and wrote the post above.

> so in some sense "ranking" them against each other, so you should be using the ranking loss. If, instead, you want to look at a score as an "absolute" measure of the likelihood of an edge's existence, you may want to check out the logistic loss instead.

I am looking for the  likelihood of edge existence, I am using:
`
lr=0.1
init_scale=0.1
comparator="dot"
bias=True
loss_fn="logistic"
`

with all_negs = True in my only relation definition as suggested in docs when sampling.

> we've seen users trying to use PBG for very small graphs
Could my issues stems from that? My training dataset is 20k edges, around 1.5k entities (vertices).

> higher dimension will capture more information and may overfit on graphs that don't hold enough information
I tried 25, 50, 100, 150 and 200 dimensions. 50 worked best for me.

> post your config so that we can see if anything looks odd there.
I hope what I posted above was enough. Again thank you for your help I really appreciate that.  
I'd note that the scores I'm seeing are (for example) 21.77.., 19.67.., I am still not sure if those are meaningful or I should just argmax to find the most likely edge.

**Comment by lerks** [comment]
My impression is that training is actually working quite well: you're getting a MRR of 0.89, which is way higher than what we get on FB15k (0.79); 81% of your test edges are the most likely edge among the ones obtained by corrupting the right-hand side entity (this is a good metric if you're interested in answering (lhs, rel, ?) queries, which may not be your case) and 99% of them rank among the top 5 (which, since you used all_negs, means 5 out of 1.5k). You AUC is also very high. Overall, I wouldn't be disappointed in such results.

One issue you might be facing in your (manual) evaluation is that you're using bias, which changes how the comparation is performed. You didn't post a full snippet, but I suspect that your `comparator` object is an instance of `DotComparator`. However, when using bias, you should use `comparator = BiasedComparator(DotComparator())`. Apart from this, I don't see anything odd in your setup.

Instead of the edge score, if you want something more immediately interpretable, you could consider passing the score through the logistic function so that you get out a probability, which is the prediction of that edge existing.

**Comment by idanh** [comment]
@lerks Thank you very much for looking at the results and configuration. Indeed was using a DotComparator without Biased wrapper.
My task at hand is link prediction, but can you point me to where I'd go about doing queries like the one you suggested (lhs, rel, ?) ?

I'll see how I can contribute to the documentation to reflect our helpful discussion here.

**Comment by lerks** [comment]
You can find example code for that in our docs, here: https://torchbiggraph.readthedocs.io/en/latest/downstream_tasks.html#ranking

**Comment by idanh** [comment]
@lerks Thank you very much for this package and your help. I'll close this issue now. Hope you have a nice day!
